From 6d76aafc74228cd986956f8152c49485b76d4b44 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:09:33 -0400
Subject: [virt] kvm/ppc: book3s hv - Handle new LPCR bits on POWER8

Message-id: <1410545655-205645-345-git-send-email-dzickus@redhat.com>
Patchwork-id: 94495
O-Subject: [RHEL7 PATCH 344/626] KVM: PPC: Book3S HV: Handle new LPCR bits on POWER8
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit e0622bd9f2fccc8a801fa7aaf4fa6d7c728c3a78
Author: Paul Mackerras <paulus@samba.org>
Date:   Wed Jan 8 21:25:27 2014 +1100

    KVM: PPC: Book3S HV: Handle new LPCR bits on POWER8

    POWER8 has a bit in the LPCR to enable or disable the PURR and SPURR
    registers to count when in the guest.  Set this bit.

    POWER8 has a field in the LPCR called AIL (Alternate Interrupt Location)
    which is used to enable relocation-on interrupts.  Allow userspace to
    set this field.

    Signed-off-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/reg.h b/arch/powerpc/include/asm/reg.h
index 419c54c..b94a440 100644
--- a/arch/powerpc/include/asm/reg.h
+++ b/arch/powerpc/include/asm/reg.h
@@ -306,8 +306,10 @@
 #define   LPCR_RMLS    0x1C000000      /* impl dependent rmo limit sel */
 #define   LPCR_RMLS_SH (63-37)
 #define   LPCR_ILE     0x02000000      /* !HV irqs set MSR:LE */
+#define   LPCR_AIL 0x01800000 /* Alternate interrupt location */
 #define   LPCR_AIL_0 0x00000000 /* MMU off exception offset 0x0 */
 #define   LPCR_AIL_3 0x01800000 /* MMU on exception offset 0xc00...4xxx */
+#define   LPCR_ONL 0x00040000 /* online - PURR/SPURR count */
 #define   LPCR_PECE 0x0001f000 /* powersave exit cause enable */
 #define     LPCR_PECEDP 0x00010000 /* directed priv dbells cause exit */
 #define     LPCR_PECEDH 0x00008000 /* directed hyp dbells cause exit */
diff --git a/arch/powerpc/kvm/book3s_hv.c b/arch/powerpc/kvm/book3s_hv.c
index 1045821..acceaa8 100644
--- a/arch/powerpc/kvm/book3s_hv.c
+++ b/arch/powerpc/kvm/book3s_hv.c
@@ -789,8 +789,11 @@ static void kvmppc_set_lpcr(struct kvm_vcpu *vcpu, u64 new_lpcr)
  /*
   * Userspace can only modify DPFD (default prefetch depth),
   * ILE (interrupt little-endian) and TC (translation control).
+  * On POWER8 userspace can also modify AIL (alt. interrupt loc.)
   */
  mask = LPCR_DPFD | LPCR_ILE | LPCR_TC;
+ if (cpu_has_feature(CPU_FTR_ARCH_207S))
+  mask |= LPCR_AIL;
  vc->lpcr = (vc->lpcr & ~mask) | (new_lpcr & mask);
  spin_unlock(&vc->lock);
 }
@@ -2166,6 +2169,9 @@ static int kvmppc_core_init_vm_hv(struct kvm *kvm)
    LPCR_VPM0 | LPCR_VPM1;
   kvm->arch.vrma_slb_v = SLB_VSID_B_1T |
    (VRMA_VSID << SLB_VSID_SHIFT_1T);
+  /* On POWER8 turn on online bit to enable PURR/SPURR */
+  if (cpu_has_feature(CPU_FTR_ARCH_207S))
+   lpcr |= LPCR_ONL;
  }
  kvm->arch.lpcr = lpcr;
 
-- 
1.7.1